### PR TITLE
sets support

### LINKED ISSUE #178: uncorrect resumtion token
Tried to update on of our servers to invenio version 3.2.
Now I can not `ListRecords`amy more. I get following error:

```
  File "/XXX/src/invenio-oaiharvester/invenio_oaiharvester/cli.py", line 87, in harvest
    setspecs
  File "/XXX/src/invenio-oaiharvester/invenio_oaiharvester/api.py", line 103, in list_records
    for record in request.ListRecords(**params):
  File "/XXX/lib/python3.6/site-packages/sickle/iterator.py", line 52, in __next__
    return self.next()
  File "/XXX/lib/python3.6/site-packages/sickle/iterator.py", line 151, in next
    self._next_response()
  File "/XXX/lib/python3.6/site-packages/sickle/iterator.py", line 138, in _next_response
    super(OAIItemIterator, self)._next_response()
  File "/XXX/lib/python3.6/site-packages/sickle/iterator.py", line 84, in _next_response
    self.oai_response = self.sickle.harvest(**params)
  File "/XXX/lib/python3.6/site-packages/sickle/app.py", line 120, in harvest
    http_response.raise_for_status()
  File "/XXX/lib/python3.6/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: UNPROCESSABLE ENTITY for url: http://localhost/oai2d?resumptionToken=.eJyNjU9vgjAchr_L70wWiiJ_bjJTiJkECC2Mi6mlKKEM0pJBNH73eZgel723N3nyPDfQQtTgm2_OxrVdz7HNtWO7G2TAyM4CfMuAbmbqrMG_QS8mVrOJJUo07QI-9ExxC4EBjRr6xx-F6lut228BdwM0V4OUx_bhh90XRlW5t2l4kaeuknUYD0G6_V1TYBkn9FqTE8YBuVbjobwsOdpOothjHvEnGKGlynPEVzRcspQeLJbjNSde-lHKNiuD1RMMvK4gplfGKNOE4Dzuu0VEOCHvyP7s6FzPrzb7b9sK_m7D_QfDQWlb.XkkFkw.ldY9EVX1H7W3LOgZv2goXLJ4M0Y&verb=ListRecords
```

During normal installation the version 1.1.1 of invenio-oaiserver gets installed. Acording to the logs of git there should be a version 1.1.2.

With version 1.0.3 of invenio-oaiserver every thing works as expected.

### PR DESCRIPTION
**Pull request by rekt-hard**
adds sets support

closes #178 , #186 ,  #191, #210

### CODE REVIEW

**Review by jma — APPROVED** [review]

File: invenio_oaiserver/config.py
Comment (ppanero) [inline invenio_oaiserver/config.py:5]: should this be a range?
```suggestion
# Copyright (C) 2021-2022 Graz University of Technology.
```

File: invenio_oaiserver/percolator.py
Comment (ppanero) [inline invenio_oaiserver/percolator.py:49]: we should try to avoid styling changes... in any case it makes me thing... If we will release this as par of a new minor version (i.e. not getting in invenio < 3.4)... can we remove Elasticsearch 5?

File: invenio_oaiserver/percolator.py
Comment (ppanero) [inline invenio_oaiserver/percolator.py:70]: same, mostly for the whole file... as much as I am inclined to use formatters... this messes with git history and might be problematic to trace changes in the future... seems to happen in the whole file...

File: invenio_oaiserver/percolator.py
Comment (ppanero) [inline invenio_oaiserver/percolator.py:133]: should this be a `current_app.logger.warning` so it is sent to Sentry?

**Review by ppanero — COMMENTED** [review]
At first glance it LGTM, I am going to test it now... The only thing where I slightly disagree are all the styling changes... as much as I would love to use formatting.... I think is good for new files but we should leave old files untouched until we manage to do it organization wide (keeping the history)....

**Comment by ppanero** [comment]
IT WORKS 🚀!

- Will require https://github.com/inveniosoftware/invenio-app-rdm/pull/1189 to work on RDM instances
- Triggers https://github.com/inveniosoftware/invenio-oaiserver/issues/210, and there are many `TODO`s in the code that would - need to be made into issues or addressed.


**List of tasks to complete before merge**:
- [ ] Take a decision on styling changes (do we revert? do we merge them?)
- [ ] Remove commented code and make into issues


**Screenshots**

List sets:

Screenshot 2022-01-31 at 09 39 28

No records: 

Screenshot 2022-01-31 at 10 11 48

Set with records (e.g. `search_pattern=metadata.description:seat`):

Screenshot 2022-01-31 at 10 10 12

**Comment by ppanero** [comment]
Removed Python 2 support 🚀

File: invenio_oaiserver/percolator.py
Comment (ppanero) [inline invenio_oaiserver/percolator.py:137]: Why do we `create` again? the inner code of the function basically creates it if doesnt exists... seems to simply be there to avoid a potential "cannot delete what's not there" error... but that should be taken care of by the `404` below.

File: invenio_oaiserver/percolator.py
Comment (ppanero) [inline invenio_oaiserver/percolator.py:130]: After talking with @rekt-hard we have deemed this line unnecesary... any ideas why it was there @slint @lnielsen

File: invenio_oaiserver/models.py
Comment (ppanero) [reply invenio_oaiserver/models.py:78]: created issue https://github.com/inveniosoftware/invenio-oaiserver/issues/211 will remove commented code.

File: tests/test_percolator.py
Comment (ppanero) [reply tests/test_percolator.py]: Im bringin back the tests...

File: invenio_oaiserver/percolator.py
Comment (ppanero) [reply invenio_oaiserver/percolator.py:44]: not worth spending the time now.

File: invenio_oaiserver/percolator.py
Comment (ppanero) [reply invenio_oaiserver/percolator.py:65]: should be avoided in the future but not worth the time now

File: tests/test_percolator.py
Comment (ppanero) [inline tests/test_percolator.py:43]: The previous tests were hard to read and I think had multiple checks for the same thing. Plus they make use of `add_record` and `remove_record` for manually managed sets, which we cannot do now with the dynamic ones.
I tried to bring back the tests that make sense and split them by use case.

File: invenio_oaiserver/fetchers.py
Comment (jma) [inline invenio_oaiserver/fetchers.py:46]: What's happens if the `search_pattern` is empty? Why the default behaviour is not to search in the `_oai.set` fields?

**Review by jma — DISMISSED** [review]

File: invenio_oaiserver/fetchers.py
Comment (jma) [reply invenio_oaiserver/fetchers.py:46]: Thanks for your reply.

Here is some suggestions:

- [ ] If the `search_pattern` is required, I suggest to add a no null constraint in the db model
- [ ] to speed up the `ListIdentifiers`, you can use `source` for the elasticsearch during the `get_records`
- [ ] the source fields can be a configuration for `ListRecords`
- [ ] using a `search_query` as `insitution:test` raise the following error: `TypeError: expected string or bytes-like object`, replacing by `institution:"test"` works (bug)

Thanks

Comment (jma) [reply invenio_oaiserver/fetchers.py:46]: > The idea is to not store the information of the sets a record belongs to within the record itself. Instead, the record is matched against the queries of the sets (ElasticSearch percolate query) and this information is generated dynamically on request.

Probably I misunderstand something. If during the request we use the `search_pattern`, is the percolator still useful?

Comment (rekt-hard) [reply invenio_oaiserver/fetchers.py:46]: > Probably I misunderstand something. If during the request we use the `search_pattern`, is the percolator still useful?

The percolate query is used to get *all* sets a record belongs to (needed for the header of the response).
The additional query above is used, when a setSpec is specified in the request (f.e. https://127.0.0.1:5000/oai2d?verb=ListRecords&metadataPrefix=oai_dc&set=my-set). It is used during record querying so only records belonging to this set are retrieved. Then we take these records and do a percolate query to get all the sets the records belong to.

Comment (lnielsen) [reply invenio_oaiserver/fetchers.py:46]: @jma First of all thanks for reviewing! 

As @rekt-hard explained, the big change here is that we want to get away from storing OAI sets information inside the record. It's highly very inefficient to do it this way (e.g. define a new set, or modify an existing set and you may need to reindex every single record in the entire repository. The only thing we're loosing is the ability to say that a record was deleted from a set. A harvest can obtain that information by running ListIdentifiers for a set and compare against a local set.

On a side note, can I ask you not to use the "Request changes" when reviewing a PR. Instead, simply prefix your comment with Comment/Question/Minor/<nothing>/Major according (see definitions [here](https://invenio.readthedocs.io/en/latest/community/contributing/contribution-guide.html#pull-requests)). Basically idea is that lack of a green approve, means it's not approved. Just wanted to explain why I'm gonna press the "dismiss review"

**Comment by slint** [comment]
Will backport some of the fixes and release in `1.3.1`.

**Pull request merged**
